### ISSUE TITLE
Unable to Loop thru prefix-list in hosts_var.

### ISSUE THREAD

**Issue by rozmusjs** [ansible-collections/cisco.ios#333]
##### SUMMARY
Unable to set prefix-list condition using loops.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
module: cisco.ios.ios_bgp_address_family
parameters: prefix_list

##### ANSIBLE VERSION

```paste below

ansible 2.10.8
  config file = /home/user1/ansible-network/ansible.cfg
  configured module search path = ['/home/user1/ansible-network/library']
  ansible python module location = /home/user1/.local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Aug 18 2020, 08:33:21) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
```

##### CONFIGURATION

```paste below
COLLECTIONS_PATHS(/home/user1/ansible-network/ansible.cfg) = ['/home/user1/ansible-network/collections']
DEFAULT_ACTION_PLUGIN_PATH(/home/user1/ansible-network/ansible.cfg) = ['/usr/local/lib/python3.6/site-packages/ara/plugins/action']
DEFAULT_CALLBACK_PLUGIN_PATH(/home/user1/ansible-network/ansible.cfg) = ['/usr/local/lib/python3.6/site-packages/ara/plugins/callback']
DEFAULT_FORKS(/home/user1/ansible-network/ansible.cfg) = 100
DEFAULT_HOST_LIST(/home/user1/ansible-network/ansible.cfg) = ['/home/user1/ansible-network/inventories']
DEFAULT_MODULE_PATH(/home/user1/ansible-network/ansible.cfg) = ['/home/user1/ansible-network/library']
DEFAULT_ROLES_PATH(/home/user1/ansible-network/ansible.cfg) = ['/home/user1/ansible-network/roles']
DEFAULT_SCP_IF_SSH(/home/user1/ansible-network/ansible.cfg) = True
GALAXY_SERVER_LIST(/home/user1/ansible-network/ansible.cfg) = ['release_galaxy']
HOST_KEY_CHECKING(/home/user1/ansible-network/ansible.cfg) = False
INTERPRETER_PYTHON(/home/user1/ansible-network/ansible.cfg) = auto
PERSISTENT_COMMAND_TIMEOUT(/home/user1/ansible-network/ansible.cfg) = 60000000
PERSISTENT_CONNECT_TIMEOUT(/home/user1/ansible-network/ansible.cfg) = 6000
```

##### OS / ENVIRONMENT

ISR4221/K9 
16.09.04

##### STEPS TO REPRODUCE

```yaml

- name: Configure BGP Address Family Settings
  cisco.ios.ios_bgp_address_family:
    config:
      as_number: "{{ item.as_number | default('123', true) }}"
      address_family:
        - afi: ipv4
          safi: unicast          
          neighbor:
            - address: "{{ item.remote_ipv4 }}"
              prefix_list:
                name: "{{ item.prefix_out | default(omit) }}"
                out: true
              soft_reconfiguration: true
    state: merged
  when:
    - item.prefix_out is defined
    - inventory_hostname in groups["ios_routers_ts"]
  with_items:
    - "{{ bgp_neighbors }}"
  tags:
    - ios-bgp
   
The relevant portion of the hosts_var:
bgp_neighbors:
  - { remote_as: "100", remote_ipv4: "1.1.1.1", des: TO_ISP,          prefix_out: TO_EXTERNAL, passwd: somepass, pwtype: 3des }
  - { remote_as: "200", remote_ipv4: "2.2.2.2", des: TO_INTERNAL prefix_out: TO_INTERNAL }

```

##### EXPECTED RESULTS

neighbor 1.1.1.1 prefix-list TO_EXTERNAL out

##### ACTUAL RESULTS

```paste below
TASK [ios-bgp : Configure BGP Address Family Settings] ***********************************************************************************************************************************************
task path: /home/user1/ansible-network/roles/ios-bgp/tasks/main.yml:40
Trying secret <ansible.parsing.vault.PromptVaultSecret object at 0x7f7021430550> for vault_id=default
Trying secret <ansible.parsing.vault.PromptVaultSecret object at 0x7f7021430550> for vault_id=default
redirecting (type: connection) ansible.builtin.network_cli to ansible.netcommon.network_cli
redirecting (type: terminal) ansible.builtin.ios to cisco.ios.ios
redirecting (type: cliconf) ansible.builtin.ios to cisco.ios.ios
<router-1> attempting to start connection
<router-1> using connection plugin ansible.netcommon.network_cli
Found ansible-connection at path /usr/local/bin/ansible-connection
<router-1> found existing local domain socket, using it!
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<router-1> updating play_context for connection
<router-1> 
<router-1> local domain socket path is /home/user1/.ansible/pc/61f599b354
Trying secret <ansible.parsing.vault.PromptVaultSecret object at 0x7f7021430550> for vault_id=default
<router-1> ANSIBLE_NETWORK_IMPORT_MODULES: disabled
<router-1> ANSIBLE_NETWORK_IMPORT_MODULES: module execution time may be extended
<router-1> ESTABLISH LOCAL CONNECTION FOR USER: user1
<router-1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o `"&& mkdir "` echo /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431 `" && echo ansible-tmp-1621272305.7401104-1243127-206140795871431="` echo /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431 `" ) && sleep 0'
Using module file /home/user1/ansible-network/collections/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py
<router-1> PUT /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/tmp_ppiy0ln TO /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py
<router-1> EXEC /bin/sh -c 'chmod u+x /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/ /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py && sleep 0'
<router-1> EXEC /bin/sh -c '/usr/libexec/platform-python /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py && sleep 0'
<router-1> EXEC /bin/sh -c 'rm -f -r /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py", line 102, in <module>
    _ansiballz_main()
  File "/home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_bgp_address_family', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py", line 2152, in <module>
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py", line 2147, in main
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/config/bgp_address_family/bgp_address_family.py", line 68, in execute_module
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/rm_base/resource_module.py", line 179, in run_commands
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/connection.py", line 195, in __rpc__
ansible.module_utils.connection.ConnectionError: neighbor 1.1.1.1 prefix_list TO_EXTERNAL out
neighbor 1.1.1.1 prefix_list TO_EXTERNAL out
                       ^
% Invalid input detected at '^' marker.

router-1(config-router-af)#
failed: [router-1] (item={'remote_as': '100', 'remote_ipv4': '1.1.1.1', 'des': 'TO_ISP', 'prefix_out': 'TO_EXTERNAL', 'passwd': 'somepass', 'pwtype': '3des'}) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": {
        "des": "TO_ISP",
        "passwd": "somepass",
        "prefix_out": "TO_EXTERNAL",
        "pwtype": "3des",
        "remote_as": "100",
        "remote_ipv4": "1.1.1.1"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272305.7401104-1243127-206140795871431/AnsiballZ_ios_bgp_address_family.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_bgp_address_family', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py\", line 2152, in <module>\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py\", line 2147, in main\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/config/bgp_address_family/bgp_address_family.py\", line 68, in execute_module\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/rm_base/resource_module.py\", line 179, in run_commands\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_ngrmr_iw/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/connection.py\", line 195, in __rpc__\nansible.module_utils.connection.ConnectionError: neighbor 1.1.1.1 prefix_list TO_EXTERNAL out\r\nneighbor 1.1.1.1 prefix_list TO_EXTERNAL out\r\n                       ^\r\n% Invalid input detected at '^' marker.\r\n\r\nrouter-1(config-router-af)#\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
Trying secret <ansible.parsing.vault.PromptVaultSecret object at 0x7f7021430550> for vault_id=default
<router-1> ANSIBLE_NETWORK_IMPORT_MODULES: disabled
<router-1> ANSIBLE_NETWORK_IMPORT_MODULES: module execution time may be extended
<router-1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o `"&& mkdir "` echo /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272308.6388416-1243127-75213335918085 `" && echo ansible-tmp-1621272308.6388416-1243127-75213335918085="` echo /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272308.6388416-1243127-75213335918085 `" ) && sleep 0'
Using module file /home/user1/ansible-network/collections/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py
<router-1> PUT /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/tmptw_9dfno TO /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272308.6388416-1243127-75213335918085/AnsiballZ_ios_bgp_address_family.py
<router-1> EXEC /bin/sh -c 'chmod u+x /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272308.6388416-1243127-75213335918085/ /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272308.6388416-1243127-75213335918085/AnsiballZ_ios_bgp_address_family.py && sleep 0'
<router-1> EXEC /bin/sh -c '/usr/libexec/platform-python /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272308.6388416-1243127-75213335918085/AnsiballZ_ios_bgp_address_family.py && sleep 0'
<router-1> EXEC /bin/sh -c 'rm -f -r /home/user1/.ansible/tmp/ansible-local-1242606x0gdvq9o/ansible-tmp-1621272308.6388416-1243127-75213335918085/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_6tzdqdam/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/facts/facts.py", line 134, in get_network_resources_facts
    self._connection, self.ansible_facts, data
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_6tzdqdam/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/facts/bgp_address_family/bgp_address_family.py", line 66, in populate_facts
    data = self.get_bgp_address_family_data(connection)
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_6tzdqdam/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/facts/bgp_address_family/bgp_address_family.py", line 50, in get_bgp_address_family_data
    return connection.get("sh running-config | section ^router bgp")
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_6tzdqdam/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/connection.py", line 195, in __rpc__
    raise ConnectionError(to_text(msg, errors='surrogate_then_replace'), code=code)
failed: [router-1] (item={'remote_as': '200', 'remote_ipv4': '2.2.2.2', 'des': 'Internal', 'prefix_out': 'TO_INTERNAL'}) => {
    "ansible_loop_var": "item",
    "changed": false,
    "invocation": {
        "module_args": {
            "config": {
                "address_family": [
                    {
                        "afi": "ipv4",
                        "aggregate_address": null,
                        "auto_summary": null,
                        "bgp": null,
                        "default": null,
                        "default_information": null,
                        "default_metric": null,
                        "distance": null,
                        "neighbor": [
                            {
                                "activate": null,
                                "additional_paths": null,
                                "address": "2.2.2.2",
                                "advertise": null,
                                "advertise_map": null,
                                "advertisement_interval": null,
                                "aigp": null,
                                "allow_policy": null,
                                "allowas_in": null,
                                "as_override": null,
                                "bmp_activate": null,
                                "capability": null,
                                "cluster_id": null,
                                "default_originate": null,
                                "description": null,
                                "disable_connected_check": null,
                                "distribute_list": null,
                                "dmzlink_bw": null,
                                "ebgp_multihop": null,
                                "fall_over": null,
                                "filter_list": null,
                                "ha_mode": null,
                                "inherit": null,
                                "internal_vpn_client": null,
                                "ipv6_adddress": null,
                                "local_as": null,
                                "log_neighbor_changes": null,
                                "maximum_prefix": null,
                                "next_hop_self": null,
                                "next_hop_unchanged": null,
                                "password": null,
                                "path_attribute": null,
                                "peer_group": null,
                                "prefix_list": {
                                    "in": null,
                                    "name": "TO_INTERNAL",
                                    "out": true
                                },
                                "remote_as": null,
                                "remove_private_as": null,
                                "route_map": null,
                                "route_reflector_client": null,
                                "route_server_client": null,
                                "send_community": null,
                                "shutdown": null,
                                "slow_peer": null,
                                "soft_reconfiguration": true,
                                "soo": null,
                                "tag": null,
                                "timers": null,
                                "transport": null,
                                "ttl_security": null,
                                "unsuppress_map": null,
                                "version": null,
                                "weight": null
                            }
                        ],
                        "network": null,
                        "redistribute": null,
                        "safi": "unicast",
                        "snmp": null,
                        "table_map": null,
                        "vrf": null
                    }
                ],
                "as_number": "123"
            },
            "running_config": null,
            "state": "merged"
        }
    },
    "item": {
        "des": "Internal",
        "prefix_out": "TO_INTERNAL",
        "remote_as": "200",
        "remote_ipv4": "2.2.2.2"
    },
    "msg": "sh running-config | section ^router bgp\r\nsh running-config | section ^router bgp\r\n ^\r\n% Invalid input detected at '^' marker.\r\n\r\nrouter-1(config-router-af)#"
}
...ignoring
redirecting (type: action) cisco.ios.ios_config to cisco.ios.ios
```

**Comment by rozmusjs** [comment]
Question:
Why the script creates "neighbor 1.1.1.1 prefix_list TO_EXTERNAL out" instead "neighbor 1.1.1.1 prefix-list TO_EXTERNAL out" statments.

**Comment by justjais** [comment]
@rozmusjs Thanks for raising the issue and this issue seems to be a duplicate of issue #267. Can you verify if the respective issue is similar to the issue you're referring, also there's a WIP PR opened to resolve the behavior and should be implemented by this week.

**Comment by rozmusjs** [comment]
Sumit; I don't think so.
In the other ticket, he's able to execute the script, his issue is that his "out" prefix-list is mapped inbound.
I can't run my playbook using the loop.
I expect Ansible to push the following statement to the device:
neighbor 1.1.1.1 prefix-list TO_EXTERNAL out
Instead, the script parameter is referenced in the actual config:
"neighbor 1.1.1.1 prefix_list TO_EXTERNAL out"

**Comment by justjais** [comment]
@rozmusjs ah got the issue, I'll append the fix to the PR #271, and once merged it'll be queued for release.

**Comment by rozmusjs** [comment]
TY, please let me know when I can download the update and try it out.

**Comment by rozmusjs** [comment]
Hi, Sumit, do you have an ETA when the update will be released?

**Comment by justjais** [comment]
@rozmusjs This will be released with IOS collection June release.

**Comment by rozmusjs** [comment]
@justjais, can I ask you to take a look at how "prefix-list" is configured in your repo?
I pulled your IOS collection (https://github.com/justjais/ios.git, ios267).
This is a very basic script:
-------------------------------------
  - name: Some BGP Config
    cisco.ios.ios_bgp_address_family:
      config:
        as_number: 65000
        address_family:
          - afi: ipv4
            safi: unicast
            bgp:
            neighbor:
              - address: 2.2.2.2
                remote_as: 65001
                prefix_list:
                  name: PREFIX-OUT
                  out: true
            network:
              - address: 1.1.1.0
                mask: 255.255.255.0
      state: merged
-------------------------------------

This is the error message:
"msg": "argument prefix_list is of type <class 'dict'> found in 'config -> address_family -> neighbor'. and we were unable to convert to list: <class 'dict'> cannot be converted to a list"

**Comment by justjais** [comment]
@rozmusjs the PR logic is still under development that's why u might have seen the above error, also that's fixed now if you want to give it a try. Once the PR is merged it'll be queued for June release.

Also, with new implementation in place `prefix_list` and `route_map` params are deprecated, and new param `prefix_lists` and `route_maps` are introduced to accommodate the list of dict for both of the params. Older deprecated params will continue to work but it'll be advisable to use the newer introduced params instead, ref: [deprecated msg](https://github.com/ansible-collections/cisco.ios/pull/271/files#diff-2670ccc3d9f4e23d3fb3f5c1612767691ac1763c7f564a9852fb934cf043f3d1R627)

**Comment by rozmusjs** [comment]
@justjais, sounds good what PR can I pull to get this tested?

**Comment by justjais** [comment]
@rozmusjs you can test with PR #271 changes.

**Comment by rozmusjs** [comment]
@justjais, I've installed PR#271, but still, see the same issue.
Any chance you can confirm prefix-list configuration works?
My script with (prefix_list) statement(and output):
---------------------------------------------------
  - name: Some BGP Config
    cisco.ios.ios_bgp_address_family:
      config:
        as_number: 65000
        address_family:
          - afi: ipv4
            safi: unicast
            bgp:
            neighbor:
              - address: 2.2.2.2
                remote_as: 65001
                prefix_list:
                  name: PREFIX-OUT
                  out: true
            network:
              - address: 192.168.0
                mask: 255.255.255.0
      state: merged
---------------------------------------------------
The full traceback is:
Traceback (most recent call last):
  File "/home/jrozmus/.ansible/tmp/ansible-local-22062303ow3e3ku/ansible-tmp-1623861205.9327576-2206472-29275792455695/AnsiballZ_ios_bgp_address_family.py", line 102, in <module>
    _ansiballz_main()
  File "/home/jrozmus/.ansible/tmp/ansible-local-22062303ow3e3ku/ansible-tmp-1623861205.9327576-2206472-29275792455695/AnsiballZ_ios_bgp_address_family.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/jrozmus/.ansible/tmp/ansible-local-22062303ow3e3ku/ansible-tmp-1623861205.9327576-2206472-29275792455695/AnsiballZ_ios_bgp_address_family.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_bgp_address_family', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py", line 2195, in <module>
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py", line 2190, in main
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/config/bgp_address_family/bgp_address_family.py", line 68, in execute_module
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/rm_base/resource_module.py", line 179, in run_commands
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/connection.py", line 195, in __rpc__
ansible.module_utils.connection.ConnectionError: neighbor 2.2.2.2 prefix_list PREFIX-OUT out
neighbor 2.2.2.2 prefix_list PREFIX-OUT out
                       ^
% Invalid input detected at '^' marker.

la3-utilc001(config-router-af)#
fatal: [la3-utilc001]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/home/jrozmus/.ansible/tmp/ansible-local-22062303ow3e3ku/ansible-tmp-1623861205.9327576-2206472-29275792455695/AnsiballZ_ios_bgp_address_family.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/home/jrozmus/.ansible/tmp/ansible-local-22062303ow3e3ku/ansible-tmp-1623861205.9327576-2206472-29275792455695/AnsiballZ_ios_bgp_address_family.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/jrozmus/.ansible/tmp/ansible-local-22062303ow3e3ku/ansible-tmp-1623861205.9327576-2206472-29275792455695/AnsiballZ_ios_bgp_address_family.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.cisco.ios.plugins.modules.ios_bgp_address_family', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py\", line 2195, in <module>\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/modules/ios_bgp_address_family.py\", line 2190, in main\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/cisco/ios/plugins/module_utils/network/ios/config/bgp_address_family/bgp_address_family.py\", line 68, in execute_module\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible_collections/ansible/netcommon/plugins/module_utils/network/common/rm_base/resource_module.py\", line 179, in run_commands\n  File \"/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_n6vxh78_/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/connection.py\", line 195, in __rpc__\nansible.module_utils.connection.ConnectionError: neighbor 2.2.2.2 prefix_list PREFIX-OUT out\r\nneighbor 2.2.2.2 prefix_list PREFIX-OUT out\r\n                       ^\r\n% Invalid input detected at '^' marker.\r\n\r\nla3-utilc001(config-router-af)#\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
...ignoring
---------------------------------------------------



And the script with (prefix_lists) statement(and output):
---------------------------------------------------
  - name: Some BGP Config
    cisco.ios.ios_bgp_address_family:
      config:
        as_number: 65000
        address_family:
          - afi: ipv4
            safi: unicast
            bgp:
            neighbor:
              - address: 2.2.2.2
                remote_as: 65001
                prefix_lists:
                  name: PREFIX-OUT
                  out: true
            network:
              - address: 192.168.0
                mask: 255.255.255.0
      state: merged
    tags:
      - ios-bgp
---------------------------------------------------
The full traceback is:
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_tmp2u_6v/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/basic.py", line 1923, in _check_argument_types
    param[k] = type_checker(value, **kwargs)
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_tmp2u_6v/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/basic.py", line 1760, in _check_type_list
    return check_type_list(value)
  File "/tmp/ansible_cisco.ios.ios_bgp_address_family_payload_tmp2u_6v/ansible_cisco.ios.ios_bgp_address_family_payload.zip/ansible/module_utils/common/validation.py", line 379, in check_type_list
    raise TypeError('%s cannot be converted to a list' % type(value))
[WARNING]: The value "65000" (type int) was converted to "'65000'" (type string). If this does not look like what you expect, quote the entire value to ensure it does not change.
fatal: [la3-utilc001]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "config": {
                "address_family": [
                    {
                        "afi": "ipv4",
                        "aggregate_address": null,
                        "auto_summary": null,
                        "bgp": null,
                        "default": null,
                        "default_information": null,
                        "default_metric": null,
                        "distance": null,
                        "neighbor": [
                            {
                                "address": "2.2.2.2",
                                "prefix_lists": {
                                    "name": "PREFIX-OUT",
                                    "out": true
                                },
                                "remote_as": 65001
                            }
                        ],
                        "network": [
                            {
                                "address": "192.168.0",
                                "mask": "255.255.255.0"
                            }
                        ],
                        "redistribute": null,
                        "safi": "unicast",
                        "snmp": null,
                        "table_map": null,
                        "vrf": null
                    }
                ],
                "as_number": "65000"
            },
            "running_config": null,
            "state": "merged"
        }
    },
    "msg": "argument prefix_lists is of type <class 'dict'> found in 'config -> address_family -> neighbor'. and we were unable to convert to list: <class 'dict'> cannot be converted to a list"
}
---------------------------------------------------

Can you share a copy of your Ansible script including prefix-list that worked for you on this PR?

**Comment by rozmusjs** [comment]
@justjais, should "prefix_list(s)" in PR271 be defined as a dictionary, should it be defined as a list of dictionaries (the same route_map is)?

**Comment by justjais** [comment]
@rozmusjs I've pushed an update and it should work as expected now for deprecated `prefix_list` with play:
```
- name: Some BGP Config
  cisco.ios.ios_bgp_address_family:
    config:
      as_number: 65000
      address_family:
        - afi: ipv4
          safi: unicast
          bgp:
          neighbor:
            - address: 2.2.2.2
              remote_as: 65001
              prefix_list:
                name: PREFIX-OUT
                out: true
          network:
            - address: 192.168.1.0
              mask: 255.255.255.0
    state: merged
```
In your play you'll need to update network address for it to work as expected.

And, for new `prefix_lists` param, play should be structured as:
```
- name: Some BGP Config
  cisco.ios.ios_bgp_address_family:
    config:
      as_number: 65000
      address_family:
        - afi: ipv4
          safi: unicast
          bgp:
          neighbor:
            - address: 2.2.2.2
              remote_as: 65001
              prefix_lists:
                - name: PREFIX-OUT
                  out: true
          network:
            - address: 192.168.1.0
              mask: 255.255.255.0
    state: merged
```
where botth `route_maps` and `prefix_list` will be list of dict instead of dict.

**Comment by rozmusjs** [comment]
Can you please advise which collection (ios_xxx) I should be pulling to
test it?

**Comment by justjais** [comment]
@rozmusjs PR with the fix is merged and will be available with June release which is planned for tomorrow, so you can test the fix with the newer IOS collection available EOW.
